File: threads02/java/ex02/src/controller/FiltroController.java
package controller;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

import thread.FiltroTask;

/**
 * Classe responsavel por controlar o processamento das tarefas
 */
public class FiltroController {
   private static final int NUM_THREADS = 2;

    /**
     * método que processa as tarefas e retorna o resultado final
     * @param partes lista de partes
     * @return lista de resultados
     */
    public List<String> processar(List<List<String>> partes) {

        if (partes == null || partes.size() != NUM_THREADS) {
            throw new IllegalArgumentException("É necessário fornecer exatamente 2 partes!");
        }

        ExecutorService executor = Executors.newFixedThreadPool(NUM_THREADS);

        List<Future<List<String>>> futuros = new ArrayList<>();

        try {

            for (List<String> parte : partes) {

                FiltroTask tarefa = new FiltroTask(parte);

                futuros.add(executor.submit(tarefa));
            }

            List<String> resultadoFinal = new ArrayList<>();

            for (Future<List<String>> futuro : futuros) {
                resultadoFinal.addAll(futuro.get());
            }

            return resultadoFinal;

        } catch (InterruptedException e) {

            Thread.currentThread().interrupt();

            throw new RuntimeException("O processamento foi interrompido", e);

        } catch (ExecutionException e) {
            throw new RuntimeException("Erro durante o processamento dos dados", e.getCause());

        } finally {
            executor.shutdown();
        }
    }
}

File: threads02/csharp/ex02/Controller/FiltroController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using ex02.Tasks;

namespace ex02.Controller;

public class FiltroController
{
    private const int NumThreads = 2;

    public async Task<List<string>> ProcessarAsync(List<List<string>> partes)
    {
        if (partes == null || partes.Count != NumThreads)
        {
            throw new ArgumentException("É necessário fornecer exatamente 2 partes!");
        }

        List<Task<List<string>>> tarefas = new();

        for (int i = 0; i < partes.Count; i++)
        {
            int numeroParte = i + 1;
            FiltroTask tarefa = new(partes[i], numeroParte);
            tarefas.Add(Task.Run(() => tarefa.Executar()));
        }

        List<string>[] resultadosDasPartes = await Task.WhenAll(tarefas);

        // Agrupa todas as listas retornadas pelas tarefas em uma única lista final
        List<string> resultadoFinal = resultadosDasPartes.SelectMany(r => r).ToList();

        return resultadoFinal;
    }
}
